src/sdk/pynni/tests/test_tuner.py

```diff
@@ -17,107 +17,172 @@
+                param = {list(search_space.keys())[0]: param}
+            for k, v in param.items():
+                if k.startswith("_mutable_layer"):
+                    _, block, layer, choice = k.split("/")
+                    cand = search_space[block]["_value"][layer].get(choice)
+                    # cand could be None, e.g., optional_inputs_chosen_state
+                    if choice == "layer_choice":
+                        self.assertIn(v, cand)
+                    if choice == "optional_input_size":
+                        if isinstance(cand, int):
+                            self.assertEqual(v, cand)
+                        else:
+                            self.assertGreaterEqual(v, cand[0])
+                            self.assertLessEqual(v, cand[1])
+                    if choice == "optional_inputs":
+                        pass  # ignore for now
+                    continue
+                item = search_space[k]
+                if item["_type"] == "choice":
+                    self.assertIn(v, item["_value"])
+                if item["_type"] == "randint":
+                    self.assertIsInstance(v, int)
+                if item["_type"] == "uniform":
+                    self.assertIsInstance(v, float)
+                if item["_type"] in ("randint", "uniform", "quniform", "loguniform", "qloguniform"):
+                    self.assertGreaterEqual(v, item["_value"][0])
+                    self.assertLessEqual(v, item["_value"][1])
+                if item["_type"].startswith("q"):
+                    multiple = v / item["_value"][2]
+                    print(k, v, multiple, item)
+                    if item["_value"][0] + EPS < v < item["_value"][1] - EPS:
+                        self.assertAlmostEqual(int(round(multiple)), multiple)
+                if item["_type"] in ("qlognormal", "lognormal"):
+                    self.assertGreaterEqual(v, 0)
+                if item["_type"] == "mutable_layer":
+                    for layer_name, info in item["_value"].items():
+                        self.assertIn(v[layer_name]["chosen_layer"], item["layer_choice"])
+
+    def search_space_test_all(self, tuner_factory, supported_types=None, ignore_types=None):
+        with open(os.path.join(os.path.dirname(__file__), "assets/search_space.json"), "r") as fp:
+            search_space_all = json.load(fp)
+        if supported_types is None:
+            supported_types = ["choice", "randint", "uniform", "quniform", "loguniform", "qloguniform",
+                               "normal", "qnormal", "lognormal", "qlognormal"]
+        full_supported_search_space = dict()
+        for single in search_space_all:
+            single_keyword = single.split("_")
+            space = search_space_all[single]
+            expected_fail = not any([t in single_keyword for t in supported_types]) or "fail" in single_keyword
+            if ignore_types is not None and any([t in ignore_types for t in single_keyword]):
```

Comment:
what is the relation between `supported_types` and `ignore_types`?

Reply:
Supported types are listed in the table. They are meant to be supported and should be correct. Other than those, all the rest are "unsupported", which are expected to produce ridiculous results or throw some exceptions. However, there are certain types I can't check. For example, generate "normal" using GP Tuner returns successfully and results are fine if we check the range (-inf to +inf), but they make no sense: it's not a normal distribution.

Reply:
please add comments for `ignore_types`